Order/guide/new order properties (#52)
plentymarkets/plenty-developers-docs: docs/en-gb/modules/rest-api-guides/pages/order-data.adoc

=== References

> The order item `references` relation is similar to `orderReferences` relation. Order items can reference other order items. 

Reviewer: ```suggestion
The order item `references` relation is similar to the `orderReferences` relation. Order items can reference other order items. 
```

The available reference types are:

* `parent`: references an order item in the parent order
* `bundle`: references an order item with `typeId == 2` in the same order (only for order items with bundle components, `typeId == 3`)
* `set`: references an order item with `typeId == 13` in the same order  (only for order items with set components, `typeId == 14`)
* `order_property`: references an order item with `typeId == 1` in the same order (only for order item with `typeId == 15`)
* `reorder`: references an order item in a reorder (order with `typeId == 12`) with stock for the current order

> Let's take a bundle as example. A bundle has components and these bundle component items reference the bundle item. The `references` relation of a component item is shown below. The bundle component item has the id 501. The bundle that the component is part of has the id 500.

Reviewer: ```suggestion
Let's take a bundle as example. A bundle has components and these bundle component items reference the bundle item. The `references` relation of a component item is shown below. The bundle component item has the ID 501. The bundle that the component is part of has the ID 500.
```

    "createdAt": "2017-12-13T16:32:53+01:00",
    "updatedAt": "2017-12-13T16:32:53+01:00"
}]
----

> == Creating an order with the new order item orderPtoperties

Reviewer: ```suggestion
== Creating an order with the new order item orderProperties
```

> The new order item orderProperties are realized as order items with `typeId` 15. The can have own prices and tax rates, depending of the property definition. Also they should have one of the new order property types:

Reviewer: ```suggestion
The new order item orderProperties are realised as order items with the `typeId` 15. They can have own prices and tax rates, depending on the property definition. They also should have one of the new order property types:
```

* order property ID = 81 (mandatory)
* order property value = 82
* order property group ID = 83 (mandatory for property groups with `groupOptionIdentifier == `groupType` and `value IN ('select', 'multiselect')`)

> As an order item orderProperty belongs to a variation order item (`typeId == 1`), a reference between the orderProperty and the variation has to be created. For this, `metaData` have to be added in the variation and the orderProperty order item.

Reviewer: ```suggestion
As an order item, orderProperty belongs to a variation order item (`typeId == 1`). Therefore, a reference between the orderProperty and the variation has to be created. For this, `metaData` have to be added in the variation and the orderProperty order item.
```

> The `metaData` of the variation order item must contain a random but unique `externalHash`, so the order item can be references correctly. The `metaData` of the orderProperty order item must have an `internalReferences` entry with `internalReferences` as key and the corresponding `externalHash` value.

Reviewer: ```suggestion
The `metaData` of the variation order item has to contain a random but unique `externalHash`, so the order item can be referenced correctly. The `metaData` of the orderProperty order item has to have an `internalReferences` entry with `internalReferences` as key and the corresponding `externalHash` value.
```

The `orderItemName` should contain:

> * the property name, if the property does not belong to a group with `select` oder `multiselect` as `groupType`.

Reviewer: ```suggestion
* the property name if the property does not belong to a group with `select` or `multiselect` as `groupType`.
```

> * the property group name, if the property belongs to a group with `select` oder `multiselect` as `groupType`. In the case of `select` the order item property with `typeId == 82` contains the selected value(s) as a JSON string.

Reviewer: ```suggestion
* the property group name if the property belongs to a group with `select` or `multiselect` as `groupType`. In the case of `select`, the order item property with `typeId == 82` contains the selected value(s) as a JSON string.
```

            "referrerId": 1.0,
            "properties": [
                {
                    "typeId": 81,
                    "value": "138"
                },
                {
                    "typeId": 82,
                    "value": "Blue"
                },
                {
                    "typeId": 83,
                    "value": "26"
                }
            ],
            "amounts": [
                {
                    "isSystemCurrency": true,
                    "currency": "EUR",
                    "exchangeRate": 1.0,
                    "priceOriginalGross": 5.00
                }
            ],
            "metaData": {
            	"internalReferences": {
            		"order_property": "12ab34cd"
            	}
            }
        },
        {
            "typeId": 15,
            "quantity": 1.0,
            "orderItemName": "Engraving",
            "countryVatId": 12,
            "vatField": 0,
            "vatRate": 19.0,
            "referrerId": 1.0,
            "properties": [
                {
                    "typeId": 81,
                    "value": "140"
                },
                {
                    "typeId": 82,

>                     "value": "For my Love ♥"

Reviewer 2: Ob es so eine gute Idee ist, den Kunden beizubringen Unicode in die Properties zu schreiben 😆  Ich will nicht wissen wo im Code das kaputt geht

Author: > Außerdem hatten wir das schon vor einiger zeit mit dem Zeichen „:heart:“. Das wird auch gerne verwendet. Daraufhin hatte ich den Zeichensatz auf „unicode“ geändert, was dann auch erstmal funktioniert hatte.

Auszug aus https://forum.plentymarkets.com/t/zeichen-werden-auf-dem-lieferschein-nicht-dargestellt/636126

Es kann also durchaus passieren, vor allem bei Kundeneingaben.
